code/game/camera_manager/camera_manager.dm

```diff
@@ -0,0 +1,224 @@
+#define RENDER_MODE_AREA 2
+
+/datum/component/camera_manager
+	var/map_name
+	var/obj/structure/machinery/camera/current
+	var/datum/shape/rectangle/current_area
+	var/atom/movable/screen/map_view/cam_screen
+	var/atom/movable/screen/background/cam_background
+	var/list/range_turfs = list()
+	/// The turf where the camera was last updated.
+	var/turf/last_camera_turf
+	var/target_x
+	var/target_y
+	var/target_z
+	var/target_width
+	var/target_height
+	var/list/cam_plane_masters
+	var/isXRay = FALSE
+	var/render_mode = RENDER_MODE_TARGET
+
+/datum/component/camera_manager/Initialize()
+	. = ..()
+	map_name = "camera_manager_[REF(src)]_map"
+	cam_screen = new
+	cam_screen.name = "screen"
+	cam_screen.assigned_map = map_name
+	cam_screen.del_on_map_removal = FALSE
+	cam_screen.screen_loc = "[map_name]:1,1"
+	cam_background = new
+	cam_background.assigned_map = map_name
+	cam_background.del_on_map_removal = FALSE
+
+	cam_plane_masters = list()
+	for(var/plane in subtypesof(/atom/movable/screen/plane_master) - /atom/movable/screen/plane_master/blackness)
+		var/atom/movable/screen/plane_master/instance = new plane()
+		add_plane(instance)
+
+/datum/component/camera_manager/Destroy(force, ...)
+	. = ..()
+	range_turfs = null
+	current_area = null
+	cam_plane_masters = null
+	QDEL_NULL(cam_background)
+	QDEL_NULL(cam_screen)
+
+/datum/component/camera_manager/proc/add_plane(atom/movable/screen/plane_master/instance)
+	instance.assigned_map = map_name
+	instance.del_on_map_removal = FALSE
+	if(instance.blend_mode_override)
+		instance.blend_mode = instance.blend_mode_override
+	instance.screen_loc = "[map_name]:CENTER"
+	cam_plane_masters["[instance.plane]"] = instance
+
+/datum/component/camera_manager/proc/register(source, mob/user)
+	var/client/user_client = user.client
+	if(!user_client)
+		return
+	user_client.register_map_obj(cam_background)
+	user_client.register_map_obj(cam_screen)
+	for(var/plane_id in cam_plane_masters)
+		user_client.register_map_obj(cam_plane_masters[plane_id])
+
+/datum/component/camera_manager/proc/unregister(source, mob/user)
+	var/client/user_client = user.client
+	if(!user_client)
+		return
+	user_client.clear_map(cam_background)
+	user_client.clear_map(cam_screen)
+	for(var/plane_id in cam_plane_masters)
+		user_client.clear_map(cam_plane_masters[plane_id])
+
+/datum/component/camera_manager/RegisterWithParent()
+	. = ..()
+	START_PROCESSING(SSdcs, src)
+	SEND_SIGNAL(parent, COMSIG_CAMERA_MAPNAME_ASSIGNED, map_name)
+	RegisterSignal(parent, COMSIG_CAMERA_REGISTER_UI, PROC_REF(register))
+	RegisterSignal(parent, COMSIG_CAMERA_UNREGISTER_UI, PROC_REF(unregister))
+	RegisterSignal(parent, COMSIG_CAMERA_SET_NVG, PROC_REF(enable_nvg))
+	RegisterSignal(parent, COMSIG_CAMERA_CLEAR_NVG, PROC_REF(disable_nvg))
+	RegisterSignal(parent, COMSIG_CAMERA_SET_AREA, PROC_REF(set_camera_rect))
+	RegisterSignal(parent, COMSIG_CAMERA_SET_TARGET, PROC_REF(set_camera))
+	RegisterSignal(parent, COMSIG_CAMERA_CLEAR, PROC_REF(clear_camera))
```

Comment:
I also noticed that these procs (including `camera_mapname_update()` from `COMSIG_CAMERA_MAPNAME_ASSIGNED`) are missing `SIGNAL_HANDLER`s.

Reply:
Good spot - thanks!